airbyte-integrations/connectors/source-amazon-ads/source_amazon_ads/streams/report_streams/report_streams.py

```diff
@@ -0,0 +1,302 @@
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in all
+# copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+#
+
+import json
+import time
+from abc import ABC, abstractmethod
+from dataclasses import dataclass
+from datetime import datetime, timedelta
+from enum import Enum
+from gzip import decompress
+from http import HTTPStatus
+from typing import Any, Dict, Iterable, List, Mapping, Optional, Tuple
+from urllib.parse import urljoin
+
+import backoff
+import pytz
+import requests
+from airbyte_cdk.logger import AirbyteLogger
+from airbyte_cdk.models import SyncMode
+from airbyte_cdk.sources.streams.http.auth import Oauth2Authenticator
+from pydantic import BaseModel
+from source_amazon_ads.common import SourceContext
+from source_amazon_ads.schemas import JSModel, MetricsReport
+from source_amazon_ads.spec import Spec
+from source_amazon_ads.streams.common import BasicAmazonAdsStream
+
+logger = AirbyteLogger()
+
+
+class RecordType(str, Enum):
+    CAMPAIGNS = "campaigns"
+    ADGROUPS = "adGroups"
+    PRODUCTADS = "productAds"
+    TARGETS = "targets"
+    ASINS = "asins"
+
+
+class Status(str, Enum):
+    IN_PROGRESS = "IN_PROGRESS"
+    SUCCESS = "SUCCESS"
+    FAILURE = "FAILURE"
+
+
+class ReportInitResponse(BaseModel):
+    reportId: str
+    status: str
+
+
+class ReportStatus(BaseModel):
+    status: str
+    location: Optional[str]
+
+
+@dataclass
+class ReportInfo:
+    report_id: str
+    profile_id: int
+    record_type: str
+
+
+class TooManyRequests(Exception):
+    """
+    Custom exception occured when response with 429 status code received
+    """
+
+
+class ReportStream(BasicAmazonAdsStream, ABC):
+    """
+    Common base class for report streams
+    """
+
+    primary_key = None
+    CHECK_INTERVAL_SECONDS = 30
+    REPORT_WAIT_TIMEOUT = timedelta(hours=1)
+    REPORT_DATE_FORMAT = "%Y%m%d"
+    cursor_field = "reportDate"
+
+    def __init__(self, config: Spec, context: SourceContext, authenticator: Oauth2Authenticator):
+        self._authenticator = authenticator
+        self._session = requests.Session()
+        self._generate_model()
+        super().__init__(config, context)
+
+    @property
+    def model(self) -> JSModel:
```

Comment:
btw, have no idea what JSModel means, more specific name? WDYT?

Reply:
Renamed it to CatalogModel